SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp

```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
+template <typename Functor>
+TFunction<void(AActor*)> RebindActorChannel(USpatialNetDriver* NetDriver, Functor&& InFunctor)
+{
+	return [NetDriver, InFunctor](AActor* Actor)
+	{
+		if (USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(Actor))
+		{
+			InFunctor(Actor, Channel);
+		}
+	};
+}
+
+void USpatialActorChannel::GetLatestAuthorityChangeFromHierarchy(USpatialNetDriver* NetDriver, const AActor* RootActor, uint64& OutTimestamp)
 {
-	if (HierarchyActor->GetIsReplicated())
+	SpatialGDK::ForeachReplicatedActorInHierarchy(const_cast<AActor*>(RootActor), RebindActorChannel(NetDriver, [&OutTimestamp] (AActor*, USpatialActorChannel* Channel)
 	{
-		if (USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(const_cast<AActor*>(HierarchyActor)))
+		if (Channel->AuthorityReceivedTimestamp > OutTimestamp)
 		{
-			if (Channel->AuthorityReceivedTimestamp > OutTimestamp)
+			OutTimestamp = Channel->AuthorityReceivedTimestamp;
+		}
+	}));
+}
+
+void USpatialActorChannel::MarkActorHierarchyForMigration(USpatialNetDriver* NetDriver, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [NetDriver](AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if (ChildActor->HasAuthority())
+		{
+			ChildActor->ForceNetUpdate();
+			Channel->MigrationVirtualWorkerId.Emplace(SpatialConstants::INVALID_VIRTUAL_WORKER_ID);
+			Channel->MigrationInitiatedFrame = NetDriver->ReplicationFrame;
+		}
+	}));
+}
+
+void USpatialActorChannel::SetActorHierarchyMigrationDestination(USpatialNetDriver* NetDriver, VirtualWorkerId NewAuthVirtualWorkerId, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [NewAuthVirtualWorkerId] (AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if(ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId = NewAuthVirtualWorkerId;
+		}
+	}));
+}
+
+void USpatialActorChannel::DropActorHierarchyMigration(USpatialNetDriver* NetDriver, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [](AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if (ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId.Reset();
+		}
+	}));
+}
+
+void USpatialActorChannel::HandleActorLoadBalancing()
+{
+	// if the field is set it means we should try to migrate authority to another worker
+	if (MigrationVirtualWorkerId.IsSet())
+	{
+		AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
+
+		// If we were locked in between replication frames, drop migration attempt.
+		if (NetDriver->LockingPolicy->IsLocked(Actor))
+		{
+			DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+		}
+		// Are we in a different frame from the one we determined we should try to migrate ?
+		else if (MigrationInitiatedFrame != NetDriver->ReplicationFrame)
```

Comment:
ReplicationFrame isn't used when running with the replication graph, the rep graph has it's own counter.